src/mvba/vcbc/mod.rs

```diff
@@ -1,75 +1,325 @@
+    pub fn c_broadcast(&mut self, msg: Vec<u8>) -> Result<()> {
+        debug_assert_eq!(self.i, self.tag.j);
+
+        // Upon receiving message (ID.j.s, in, c-broadcast, m):
+        // send (ID.j.s, c-send, m) to all parties
+        let send_msg = Message {
+            tag: self.tag.clone(),
+            action: MSG_ACTION_C_SEND.to_string(),
+            m: msg,
+            sig: Sig::None,
+        };
+        self.broadcast(&send_msg)?;
+        self.decide()
+    }
 
-        self.state.set_proposal(proposal, &mut self.ctx)?;
-        if let Some(s) = self.state.decide(&mut self.ctx)? {
-            self.state = s;
+    // log_message logs (adds) messages into the message_log
+    pub fn log_message(&mut self, sender: NodeId, msg: Message) -> Result<()> {
+        if msg.tag != self.tag {
+            log::trace!("invalid tag, ignoring message.: {:?}. ", msg);
+            return Ok(());
         }
+
+        log::debug!("{:?} adding message: {:?}", self.state, msg);
+        self.message_log
+            .entry(msg.action.clone())
+            .or_default()
+            .push((sender, msg));
+
         Ok(())
     }
 
-    pub fn process_message(&mut self, sender: &NodeId, message: &[u8]) -> Result<()> {
-        let msg: Message = bincode::deserialize(message)?;
+    pub fn is_delivered(&self) -> bool {
+        self.u_bar.is_some()
+    }
+
+    // decides read messages from the message log and decides what to do based on
+    // the current state of the system.
+    pub fn decide(&mut self) -> Result<()> {
+        // First we check if the message is c-delivered or not.
+        self.check_is_delivered()?;
+
+        match self.state {
+            State::Send => {
+                // Upon receiving message (ID.j.s, c-send, m) from Pl:
+                match self.message_log.get(MSG_ACTION_C_SEND) {
+                    Some(msgs) => {
+                        // TODO: better way?
+                        let msg_cloned = msgs.clone();
+                        for (l, msg) in msg_cloned {
+                            // if j = l and m̄ = ⊥ then
+                            if l == self.tag.j && self.m_bar.is_none() {
+                                // m̄ ← m
+                                self.m_bar = Some(msg.m);
+
+                                // Memorizing the hash of message
+                                self.d = Some(Hash32::calculate(self.m_bar.as_ref().unwrap()));
+
+                                // compute an S1-signature share ν on (ID.j.s, c-ready, H(m))
+                                let sign_bytes = self.sign_bytes();
+                                let s1 = self.sec_key_share.sign(sign_bytes);
+                                let ready_msg = Message {
+                                    tag: self.tag.clone(),
+                                    action: MSG_ACTION_C_READY.to_string(),
+                                    m: self.d.as_ref().unwrap().to_bytes(),
```

Comment:
We don't allow panics in non-test code so we'll need to figure out how to do this without an unwrap

Reply:
In my point of view, better we panic here. Because here we are going to send a ready message that we Must have. Otherwise the protocol has critical problems.

Reply:
This is a team rule we follow, no `unwrap()`, `panic!()`, `expect()` etc. in non-test code.

There is almost always a way to re-design things to avoid these panics, but when we can't, we return an error i.e. `Err::InvalidState` something